integration_test/stream_test.go

```diff
@@ -285,337 +79,210 @@ func TestPubSub(t *testing.T) {
-						return
-					}
+		}).
+		asyncPublish(expectedRecords)
+
+	// setup consumer
+	testContext.AddConsumerContext(common.GenerateNodeId(), testParams.topic, testParams.fragmentOffsets).
+		onComplete(func() {
+			for _, record := range receivedRecords {
+				if !contains(expectedRecords, record) {
+					t.Errorf("Record(%s) is not exists", record)
 				}
 			}
-		}()
-
-		go func() {
-			for index, record := range records {
-				publishCh <- &client.PublishData{
-					Data:   record,
-					NodeId: nodeId,
-					SeqNum: uint64(index),
-				}
+		}).
+		onError(func(err error) {
+			t.Error(err)
+		}).
+		onSubscribe(testParams.consumerBatchSize, testParams.consumerFlushInterval, func(received *client.SubscribeResult) bool {
+			receivedRecords = append(receivedRecords, received.Items[0].Data)
+			if len(receivedRecords) == len(expectedRecords) {
+				fmt.Println("consumer is finished")
+				return true
+			} else {
+				return false
 			}
-		}()
-
-		return records
-	}
-
-	// Start producer
-	var totalPublishedRecords [][]byte
-	totalPublishedRecords = append(totalPublishedRecords, runProducer("data1.txt")...)
+		}).
+		waitFinished()
+}
 
-	// Start consumer
-	type SubscribedRecords [][]byte
-	var totalSubscribedRecords []SubscribedRecords
-	var mu sync.Mutex
+func TestMultiFragmentsOptionalConsume(t *testing.T) {
+	testContext := DefaultShapleQTestContext(t)
+	testContext.
+		WithBrokerTimeout(1500).
+		RunBrokers().
+		SetupTopics()
+	defer testContext.Terminate()
+
+	// test body
+	testParams := testContext.TestParams()
+	var expectedRecords [][]byte = testParams.testRecords[0]
+	receivedRecords := map[uint32]records{}
 	var wg sync.WaitGroup
+	var mu sync.Mutex
 
-	runConsumer := func(name string) {
-		var subscribedRecords SubscribedRecords
-
-		consumerConfig := config2.NewConsumerConfig()
-		consumerConfig.SetLogLevel(testLogLevel)
-		consumerConfig.SetServerAddresses(zkAddrs)
-		consumerConfig.SetBrokerTimeout(30000)
-		consumer := client.NewConsumer(consumerConfig, topic, fragmentOffsets)
-		if err := consumer.Connect(); err != nil {
-			t.Error(err)
-			wg.Done()
-			return
-		}
-
-		receiveCh, subErrCh, err := consumer.Subscribe(maxBatchSize, flushInterval)
-		if err != nil {
+	// setup producer
+	testContext.AddProducerContext(common.GenerateNodeId(), testParams.topic).
+		onError(func(err error) {
 			t.Error(err)
-			wg.Done()
-			return
-		}
-
-		go func() {
-			defer wg.Done()
-			defer consumer.Close()
-
-			for {
-				select {
-				case received := <-receiveCh:
-					for _, data := range received.Items {
-						subscribedRecords = append(subscribedRecords, data.Data)
-					}
-
-					if len(subscribedRecords) == len(totalPublishedRecords) {
-						fmt.Printf("done %s\n", name)
+		}).
+		asyncPublish(expectedRecords)
+
+	// setup consumer for each fragment
+	for fragmentId, offset := range testParams.fragmentOffsets {
+		wg.Add(1)
+		func(fid uint32, startOffset uint64) {
```

Comment:
Why use lambda in here? May I have leaked go knowledge?

Reply:
to capture the values inside

Reply:
It can just replaced to new variable assignment.. I'll fix it